virtool/samples/data.py

```diff
@@ -423,3 +424,55 @@ async def move_sample_files(self, progress_handler: AbstractProgressHandler):
 
             await virtool.samples.db.move_sample_files_to_pg(self._db, self._pg, sample)
             await tracker.add(1)
+
+    async def deduplicate_sample_names(self):
+        """
+        Find all samples with duplicate names in the same space and rename
+        them with increasing integers by order of creation.
+        """
+        async with self._db.create_session() as session:
+            async for duplicate_name_documents in self._db.samples.aggregate(
+                [
+                    {
+                        "$group": {
+                            "_id": {"name": "$name", "space_id": "$space_id"},
+                            "count": {"$sum": 1},
+                            "documents": {
+                                "$push": {"_id": "$_id", "created_at": "$created_at"}
+                            },
+                        }
+                    },
+                    {"$match": {"count": {"$gt": 1}}},
+                    {"$unwind": "$documents"},
+                    {"$sort": {"documents.created_at": 1}},
+                    {
+                        "$group": {
+                            "_id": "$_id",
+                            "sample_ids": {"$push": "$documents._id"},
+                        }
+                    },
+                    {
+                        "$project": {
+                            "name": "$_id.name",
+                            "space_id": "$_id.space_id",
+                            "_id": 0,
+                            "sample_ids": 1,
+                        }
+                    },
+                ],
+                session=session,
+            ):
+                name_generator = incrementing_sample_name_generator(
+                    self._db,
+                    duplicate_name_documents["name"],
+                    duplicate_name_documents.get("space_id", None),
+                    session,
+                )
+
+                for sample_id in duplicate_name_documents["sample_ids"][1:]:
+                    new_name = await name_generator.asend(None)
```

Comment:
I think it would be better to make a simple class to do this instead of an async generator.

```
name_generator = NameGenerator(db, name, space_id, session)

name = name_generator.get()
```

The nice thing about generators is being able to use `for` to iterate through them. The iterable nature of the generator isn't being used here.

Thoughts?

Reply:
Hmm yeah your probably right. This method cuts down boilerplate a bit, but it doesn't really need to be a generator. Plus the `generator.asend` syntax is pretty awkward for getting a new name. Will convert to a class